lib/graphql/schema/warden.rb

```diff
@@ -126,8 +126,53 @@ def root_type_for_operation(op_name)
 
       private
 
+      def union_memberships(obj_type)
+        @unions ||=  read_through { |obj_type| @schema.union_memberships(obj_type) }
+        @unions[obj_type]
+      end
+
       def visible_field?(field_defn)
-        visible?(field_defn) && visible?(field_defn.type.unwrap)
+        visible?(field_defn) && visible_type?(field_defn.type.unwrap)
+      end
+
+      def visible_type?(type_defn)
+        return false unless visible?(type_defn)
+
+        if type_defn.kind.object? && has_abstract_type?(type_defn)
+          visible_implementation?(type_defn)
+        else
+          referenced_by_visible_members?(type_defn)
+        end
+      end
```

Comment:
I _think_ another valid implementation of this method is: 

```ruby 
# Type `t` is visible if: 
# - It's visible in its own right AND 
# - Either: 
#   - Some parts of the schema reference it OR 
#   - Some of its abstract types are visible 
def visible_type?(t) 
  visible?(t) && (
      referenced_by_visible_members?(t) || 
      member_or_implements?(t)
    )
end 
```

What do you think of that implementation? Is it valid? Are there other upsides/downsides? 

Personally I prefer it because we can remove some of intermediary methods and I find the control flow a bit easier to follow. I don't _think_ it would have much worse performance, if worse at all. What do you think?

Reply:
The problem with this one is that is would return true in the case where for example:

  - orphan type A implements interface B
  - interface B is hidden

I think type A should be hidden in that case, but currently `referenced_by_visible_members?` will return true because type A has no references (we show it in those cases)

Something that would allow use this simpler logic would be to hide always when there are no references, but that seems like a weird default that would confuse people 🤔

Reply:
That brings a question though, should types that are not part of any abstract types and not referenced by anything even be allowed?

Reply:
> `referenced_by_visible_members?` will return true

ohhh i missed that! 🙇 

> not part of any abstract types and not referenced by anything

I _think_ it's ok, i mean, from a client's point of view, what's the difference compared to now? 

A type which is not part of an abstract type can't be spread anywhere, so that's ok. 

I guess `__type(name: "...")` would start returning `nil`, but I think that's ok too. 

Any reason _not_ to do it?

Reply:
I'll update the PR to have that behaviour, I think it makes sense, just afraid it's confusing to users 🤔

Reply:
>  If none of your fields use Int, should Int be hidden?

I _think_ this is the current behavior!